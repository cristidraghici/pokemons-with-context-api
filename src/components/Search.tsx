import { useState, useRef, useEffect } from 'react'
import { useGlobalContext } from '../contexts'
import debounce from '../utils/debounce'

import { useFetchPokemon } from '../hooks'

const Search = (): JSX.Element => {
  const {
    state: { pokemonName, isLoading, error },
    dispatch,
  } = useGlobalContext()
  const [searchInput, setSearchInput] = useState(pokemonName)

  useFetchPokemon()

  const init = () => setSearchInput('pikachu')

  const debounced = useRef(
    debounce((value: string) => {
      dispatch({
        type: 'setPokemonName',
        payload: value,
      })
    }, 500)
  )
  useEffect(() => init(), [])
  useEffect(() => debounced.current(searchInput), [searchInput])

  return (
    <>
      <input
        className="Search"
        type="text"
        value={searchInput}
        onChange={(e) => {
          setSearchInput(e.target.value)
          dispatch({
            type: 'setError',
            payload: '',
          })
        }}
      />
      {isLoading && <div className="Loading">Loading...</div>}
      {!!error && <div className="Error">An error occurred while retrieving the data.</div>}
    </>
  )
}
export default Search
