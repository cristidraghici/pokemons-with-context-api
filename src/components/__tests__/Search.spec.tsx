import { render } from '@testing-library/react'
import Search from '../Search'

describe('./components/Search', () => {
  it('matches snapshot', () => {
    const { container } = render(<Search />)
    expect(container).toMatchSnapshot()
  })
})
