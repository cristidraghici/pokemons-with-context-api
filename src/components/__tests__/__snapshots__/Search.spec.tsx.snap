// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./components/Search matches snapshot 1`] = `
<div>
  <input
    class="Search"
    type="text"
    value="pikachu"
  />
</div>
`;
